lib/mint/web_socket/per_message_deflate.ex

```diff
@@ -55,6 +58,12 @@ defmodule Mint.WebSocket.PerMessageDeflate do
         :default
       )
 
+    if Keyword.has_key?(opts, :zlib_controlling_process) do
+      pid_string = Keyword.get(opts, :zlib_controlling_process)
+      :ok = :zlib.set_controlling_process(inflate_zstream, :erlang.list_to_pid(pid_string))
+      :ok = :zlib.set_controlling_process(deflate_zstream, :erlang.list_to_pid(pid_string))
```

Comment:
Why convert the pid to a list and then back? Why not just pass the pid as-is?

Reply:
```
** (Protocol.UndefinedError) protocol String.Chars not implemented for #PID<0.429.0> of type PID. This protocol is implemented for the following type(s): Atom, BitString, Date, DateTime, Decimal, Float, Hex.Solver.Assignment, Hex.Solver.Constraints.Empty, Hex.Solver.Constraints.Range, Hex.Solver.Constraints.Union, Hex.Solver.Incompatibility, Hex.Solver.PackageRange, Hex.Solver.Term, Integer, List, MyXQL.Queries, MyXQL.Query, MyXQL.TextQueries, MyXQL.TextQuery, NaiveDateTime, Time, URI, Version, Version.Requirement
    (elixir 1.14.3) lib/string/chars.ex:3: String.Chars.impl_for!/1
    (elixir 1.14.3) lib/string/chars.ex:22: String.Chars.to_string/1
    (mint_web_socket 1.0.2) lib/mint/web_socket.ex:572: anonymous fn/1 in Mint.WebSocket.normalize_params/1
    (elixir 1.14.3) lib/enum.ex:1658: Enum."-map/2-lists^map/1-0-"/2
    (mint_web_socket 1.0.2) lib/mint/web_socket.ex:570: Mint.WebSocket.normalize_params/1
    (mint_web_socket 1.0.2) lib/mint/web_socket.ex:553: anonymous fn/1 in Mint.WebSocket.get_extensions/1
    (elixir 1.14.3) lib/enum.ex:1658: Enum."-map/2-lists^map/1-0-"/2
    (mint_web_socket 1.0.2) lib/mint/web_socket.ex:260: Mint.WebSocket.do_upgrade/6
Function: #Function<9.97798924/0 in HedwigSlackSocket.WebSocketPool.init_worker/1>
    Args: []
```

I'm certainly open to better ideas!

Reply:
Ah, that's because it's being passed in `params` instead of `opts`, see: https://github.com/elixir-mint/mint_web_socket/blob/7e298781c7aa9525d63c784d78f6e62ea51825ea/lib/mint/web_socket.ex#L548-L565

`params` are sent to the remote during extension negotiation so they must be able to be concatenated into the binary for a header, see https://datatracker.ietf.org/doc/html/rfc6455#section-9.1

Instead in the calling code you should use:

```elixir
Mint.WebSocket.upgrade(:wss, conn, path, [],
  extensions: [
    {Mint.WebSocket.PerMessageDeflate,
     [],
     zlib_controlling_process: parent}
  ]
)
```

instead to pass `zlib_controlling_process` as an opt instead of a param.